### PR TITLE
Enable can match for search with reader contexts

### PR DESCRIPTION
**Pull request by dnhatn**
Enabling the can-match phase for search requests with reader contexts

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 1]
```diff
@@ -1138,17 +1138,38 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
...
-        // and we can afford false positives.
-        final boolean hasRefreshPending = indexShard.hasRefreshPending();
-        try (Engine.Searcher searcher = indexShard.acquireSearcher("can_match")) {
+        final boolean hasRefreshPending;
+        final Engine.Searcher canMatchSearcher;
+        if (request.readerId() != null) {
+            final ReaderContext readerContext = findReaderContext(request.readerId());
+            checkKeepAliveLimit(request.keepAlive().millis());
+            readerContext.keepAlive(request.keepAlive().millis());
+            readerContext.incRef();
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1148]: Why not using `markAsUsed` ?

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 2]
```diff
@@ -1138,17 +1138,38 @@ public AliasFilter buildAliasFilter(ClusterState state, String index, Set<String
...
+            } finally {
+                if (success == false) {
+                    readerContext.decRef();
+                }
+            }
+            hasRefreshPending = false;
+        } else {
+            IndexShard indexShard = indexService.getShard(request.shardId().getId());
+            hasRefreshPending = indexShard.hasRefreshPending();
+            // we don't want to use the reader wrapper since it could run costly operations
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchService.java:1165]: Unrelated but I think we can remove this comment now

**Review by jimczi — COMMENTED** [review]
Looks great, I left one comment.

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:1148]: `markAsUsed` is not water-tight. It's possible that the Reaper already decides to reap this reader context, then `markAsUsed` will be a noop, and the searcher can be closed in midway (see https://github.com/elastic/elasticsearch/pull/20095). Should we make `markAsUsed` as tighten as `incRef`?

**Comment by dnhatn** [comment]
@jimczi I pushed changes. Can you take another look? Thank you!

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks Jim!

**Pull request merged**